python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py

```diff
@@ -0,0 +1,80 @@
+import gc
+from typing import NamedTuple
+
+import objgraph
+import pytest
+
+from dagster._check import CheckError
+from dagster._utils.cached_method import cached_method
+
+
+def test_cached_method():
+    class MyClass:
+        def __init__(self, attr1):
+            self._attr1 = attr1
+            self.calls = []
+
+        @cached_method
+        def my_method(self, arg1):
+            self.calls.append(arg1)
+            return (arg1, self._attr1)
+
+    obj1 = MyClass(4)
+    assert obj1.my_method(arg1="a") == ("a", 4)
+    assert obj1.my_method(arg1="a") == ("a", 4)
+    assert obj1.my_method(arg1="b") == ("b", 4)
+    assert obj1.my_method(arg1="a") == ("a", 4)
+    assert obj1.calls == ["a", "b"]
+
+    obj2 = MyClass(5)
+    assert obj2.my_method(arg1="a") == ("a", 5)
+    assert obj2.my_method(arg1="b") == ("b", 5)
+    assert obj2.calls == ["a", "b"]
+
+
+def test_kwargs_order_irrelevant_and_no_kwargs():
+    class MyClass:
+        def __init__(self):
+            self.calls = []
+
+        @cached_method
+        def my_method(self, arg1, arg2):
+            self.calls.append(arg1)
+            return arg1, arg2
+
+    obj1 = MyClass()
+    assert obj1.my_method(arg1="a", arg2=5) == ("a", 5)
+    assert obj1.my_method(arg2=5, arg1="a") == ("a", 5)
+    assert len(obj1.calls) == 1
+
+    with pytest.raises(CheckError, match="does not support non-keyword arguments"):
+        obj1.my_method("a", 5)
+
+
+def test_does_not_leak():
+    class KeyClass(NamedTuple):
+        attr: str
+
+    class ValueClass(NamedTuple):
+        attr: str
+
+    class MyClass:
+        @cached_method
+        def my_method(self, arg1: KeyClass) -> ValueClass:
+            return ValueClass("abc")
+
+    obj = MyClass()
+    assert objgraph.count("MyClass") == 1
+    assert objgraph.count("KeyClass") == 0
+    assert objgraph.count("ValueClass") == 0
+
+    obj.my_method(arg1=KeyClass("1234"))
+    assert objgraph.count("MyClass") == 1
+    assert objgraph.count("KeyClass") == 1
```

Comment:
is the `KeyClass` instance guaranteed to stick around? the cache doesn't hold a ref right, just uses it to calc a key? Think this might flake if gc runs at different times and clears that

Reply:
The cache key includes a ref. If we didn't hold a ref, we'd need some sort of scheme for generating a key that respects `__eq__` of the key class, which I don't think is possible in the general case.